libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,269 @@
+        """Upsert a stream of data by id."""
+
+    @abc.abstractmethod
+    def get_by_ids(
+        self,
+        ids: Iterable[str],  # Sequence or Iterable
+        /,
+    ) -> Iterable[T]:
+        """Get items by id."""
+
+    @abc.abstractmethod
+    def delete_by_ids(
+        self,
+        ids: Sequence[str],  # Sequence or Iterable
+        /,
+        **kwargs: Any,
+    ) -> DeleteResponse:
+        """Delete an item by id."""
+
+    # Delete and get are part of the READ/WRITE interface.
+    # They do not take advantage of indexes on the content.
+    # However, all the indexers ARE assumed to have the capability to index
+    # on metadata if they implement the delete_by_query and get_by_query methods.
+    @abc.abstractmethod
+    def lazy_get(self, query: FilterQuery, /, **kwargs: Any) -> Iterable[T]:
+        """Get items by query."""
+
+    @abc.abstractmethod
+    def lazy_delete(
+        self, query: FilterQuery, /, **kwargs: Any
+    ) -> Iterable[DeleteResponse]:
+        """Delete items by query."""
+        # Careful with halloween problem
+
+
+class Hit(TypedDict):
+    # Should subclass from content?
+    id: str
+    score: float
+    # Optional source
+    source: Optional[T]
+    metadata: Dict[str, Any]
+
+
+class QueryResponse(TypedDict):
+    """A retrieval result."""
+
+    # Free form metadata for vectorstore providers
+    metadata: Dict[str, Any]
+    hits: List[Hit]
+
+
+class SearchQuery(FilterQuery, Generic[T]):
+    """A query that searches for items according to some criteria.
+
+    This query format is specifically adapted to vectorstores.
+
+    It supports similarity searches together with standard filtering operations.
+
+    We would need to verify that we can accommodate all the existing functionality
+    with it.
+    """
+
+    query: Optional[T]
+    # Optionally specify the method to use for similarity search
+    method: NotRequired[Optional[str]]  # e.g., "cosine", "euclidean", "jaccard", "mmr"
+
+
+class UpsertData(TypedDict):
+    """Data to upsert."""
+
+    id: str
+    document: Document
+    vector: Vector
+
+
+class VectorStore(BaseIndex[Document]):
+    @abc.abstractmethod
+    def query(
+        self,
+        query: SearchQuery[T],
+        /,
+        *,
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Any,
+    ) -> QueryResponse[
+        Document
+    ]:  # kwargs for things that are not captured via the standard query?
+        """Query for items."""
+
+    @abc.abstractmethod
+    def upsert_with_vector(
+        self,
+        data: Iterable[UpsertData],
+        /,
+        **kwargs: Any,
+    ) -> Iterable[UpsertResponse]:
+        """Upsert vectors by id."""
+
+    # --------------
+    # TODO: Discuss whether we want this
+    class ContentWithExtras(TypedDict, Generic[T]):
+        content: T
+        metadata: dict
+
+    @abc.abstractmethod
+    def get_with_extras(
```

Comment:
could query just support get by id

Reply:
Yeah that's the thing to do -- we'll need to standardize syntax for the ID field. Perhaps `$.ID`